lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -95,6 +101,109 @@ def generate_replace_result_xml(result_sourcedid, score):
     return etree.tostring(xml, xml_declaration=True, encoding='UTF-8')
 
 
+def get_assignments_for_problem(problem_descriptor, user_id, course_key):
+    """
+    Trace the parent hierarchy from a given problem to find all blocks that
+    correspond to graded assignment launches for this user. A problem may
+    show up multiple times for a given user; the problem could be embedded in
+    multiple courses (or multiple times in the same course), or the block could
+    be embedded more than once at different granularities (as an individual
+    problem and as a problem in a vertical, for example).
+
+    Returns a dict where the key is a usage descriptor for a block in the
+    parent hierarchy, and the value is a list of GradedAssignment objects that
+    are associated with that descriptor for the current user.
+    """
+    assignments = defaultdict(list)
+    current_descriptor = problem_descriptor
+    while current_descriptor:
+        assignments_for_block = GradedAssignment.objects.filter(
+            user=user_id, course_key=course_key, usage_key=current_descriptor.location
+        )
+        for assignment in assignments_for_block:
+            assignments[current_descriptor].append(assignment)
+        current_descriptor = current_descriptor.get_parent()
+    return assignments
```

Comment:
Does this dict require holding references to the descriptors, as opposed to just the locations? Locations are much smaller, trivial to serialize, less likely to cause memory leak problems down the road.